containers/src/Data/Tree.hs

```diff
@@ -181,14 +182,46 @@ mfixTree f
 instance Traversable Tree where
     traverse f (Node x ts) = liftA2 Node (f x) (traverse (traverse f) ts)
 
+-- | Folds in preorder
 instance Foldable Tree where
-    foldMap f (Node x ts) = f x `mappend` foldMap (foldMap f) ts
+    fold = foldMap id
+    {-# INLINABLE fold #-}
+
+    foldMap f = go
+      where go (Node x ts) = f x `mappend` foldMap go ts
+    {-# INLINE foldMap #-}
+
+    foldr f z = \t -> go t z  -- Use a lambda to allow inlining with two arguments
+      where
+        go (Node x ts) = f x . foldr ((.) . go) id ts
+        -- This has been found to optimize better than the more obvious definition of
+        -- go (Node x ts) z' = f x (foldr go z' ts)
+    {-# INLINE foldr #-}
+
+    foldl' f = go
+      where go !z (Node x ts) = foldl' go (f z x) ts
+    {-# INLINE foldl' #-}
+
+    foldr1 f = go
+      where go (Node x ts) = foldr (\n k prev -> f prev (k (go n))) id ts x
+
+    foldl1 f (Node x ts) = foldl (foldl f) x ts
 
     null _ = False
     {-# INLINE null #-}
 
-    toList = flatten
-    {-# INLINE toList #-}
+    elem = any . (==)
+    {-# INLINABLE elem #-}
+
+    maximum = foldl1' max
```

Comment:
You should also implement `sum` and `product`, as you on this.

And `foldMap'` for `base >=4.13`.

Reply:
As I mentioned in the description `sum`, `product` and `foldMap'` are working well by default. The only thing we gain by implementing these is saving one operation with the initial value. Which I can implement, sure, but is that what you're asking for?

Reply:
description is not visible in the source nor version control. Why `elem` is expanded, but `foldMap'` isn't? Write the note.

I also doubt that `sum` works well on old GHCs, before it was changed to use `foldMap'`.

Reply:
Done, added a note and implemented `sum` and `product`.